apps/liteprotocoltester/statistics.nim

```diff
@@ -274,16 +279,43 @@ proc jsonStats*(self: PerPeerStatistics): string =
       "{\"result:\": \"Error while generating json stats: " & getCurrentExceptionMsg() &
       "\"}"
 
-proc checkIfAllMessagesReceived*(self: PerPeerStatistics): Future[bool] {.async.} =
+proc lastMessageArrivedAt*(self: PerPeerStatistics): Result[Moment, void] =
+  var lastArrivedAt = Moment.init(0, Millisecond)
+  for stat in self.values:
+    let lastMsgFromPeerAt = stat.lastMessageArrivedAt().valueOr:
+      continue
+
+    if lastMsgFromPeerAt > lastArrivedAt:
+      lastArrivedAt = lastMsgFromPeerAt
+
+  if lastArrivedAt == Moment.init(0, Millisecond):
+    return err()
+
+  return ok(lastArrivedAt)
+
+proc checkIfAllMessagesReceived*(
+    self: PerPeerStatistics, maxWaitForLastMessage: Duration
+): Future[bool] {.async.} =
   # if there are no peers have sent messages, assume we just have started.
   if self.len == 0:
     return false
 
+  var messageCheck = true
   for stat in self.values:
     if (stat.allMessageCount == 0 and stat.receivedMessages == 0) or
         stat.helper.maxIndex < stat.allMessageCount:
+      messageCheck = false
+      break
+
+  if not messageCheck:
+    let lastMessageAt = self.lastMessageArrivedAt().valueOr:
       return false
 
+    if Moment.now() - lastMessageAt < maxWaitForLastMessage:
+      return false
+    else:
+      info "No message since max wait time", maxWait = $maxWaitForLastMessage
+
```

Comment:
I'm struggling to understand this part of the code, could you please explain it in a few words? Apologies, maybe I lack context or today I'm particularly dumb lol

Reply:
No way, it is my fault if it is not straightforward enough. Apologise.

receiver side can collect messages from many publisher at once. It creates statistics separately of them.
This part tries to get the last message arrived among all possible sender/publisher peer.

Than here even if we think not all of the messages arrived we check a time limit, a max wait before we can consider the test ended on the other parties.
So we check this threshold against last message arrival.

The issue I faced that if there were some message lost (maybe because of filter service failure) the receiver though it can expect more messages to come yet did not exit... than it screw up the dashboard statistics.

I will add comment to the code!

Reply:
Ooh now I understood it! Thanks so much! 😍